File: 백준/Silver/2531. 회전 초밥/회전 초밥.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [N, d, k, c] = input[0].split(" ").map(Number)
const sushi = input.slice(1).map(Number)
const kind = {}
let answer = Number.MIN_SAFE_INTEGER
let cnt = 0

for (let i = 0; i < k; i++) {
  if (!kind[sushi[i]]) {
    kind[sushi[i]] = 0
    cnt += 1
  }
  kind[sushi[i]] += 1
}
if (!kind[c]) {
  kind[c] = 0
  cnt += 1
}
kind[c] += 1
answer = Math.max(answer, cnt)

for (let i = 1; i < N; i++) {
  kind[sushi[i - 1]] -= 1
  if (kind[sushi[i - 1]] == 0) {
    cnt -= 1
  }
  if (!kind[sushi[(i + k - 1) % N]]) {
    kind[sushi[(i + k - 1) % N]] = 0
    cnt += 1
  }
  kind[sushi[(i + k - 1) % N]] += 1

  answer = Math.max(answer, cnt)
}

console.log(answer)

File: 백준/Silver/2531. 회전 초밥/회전 초밥.py
import sys
input = sys.stdin.readline

N,d,k,c = map(int,input().split())
sushi = [int(input()) for i in range(N)]
check = {}
cnt = 0
answer = -1e9

for i in range(0,k):
  if(check.get(sushi[i]) == None):
    check[sushi[i]] = 0
    cnt+=1
  check[sushi[i]] +=1
    
if(check.get(c) == None):
    cnt+=1
    check[c] = 0
check[c] += 1
answer = max(answer,cnt)


for i in range(1,N):
  check[sushi[i-1]] -= 1
  if(check[sushi[i-1]] == 0):
    del check[sushi[i-1]]
    cnt -=1

  if(check.get(sushi[(i+k-1)%N]) == None):
    check[sushi[(i+k-1)%N]] =0
    cnt +=1
  check[sushi[(i+k-1)%N]] +=1
  answer = max(answer,cnt)
  
print(answer)